### PR TITLE
change api to remove setting the access token explicitly

### PR DESCRIPTION
**Pull request by abhishek1508**
## Description

Change API to read the access token internally without asking the external developer to set it again. Fixes [#3341](https://github.com/mapbox/mapbox-navigation-android/issues/3341)

- [x] I have added any issue links
- [x] I have added all related labels (`bug`, `feature`, `new API(s)`, `SEMVER`, etc.)
- [x] I have added the appropriate milestone and project boards

### Goal
Change API to read the access token internally without asking the external developer to set it again. Toggle Guidance Image based on banner instruction rather than route progress.

### Implementation

Change API to read the access token internally without asking the external developer to set it again. Toggle Guidance Image based on banner instruction rather than route progress.

## Screenshots or Gifs

Please include all the media files to give some context about what's being implemented or fixed. It's not mandatory to upload screenshots or gifs, but for most of the cases it becomes really handy to get into the scope of the feature / bug being fixed and also it's REALLY useful for UI related PRs screenshot gif

## Testing

Please describe the manual tests that you ran to verify your changes

- [x] I have tested locally (including `SNAPSHOT` upstream dependencies if needed) through testapp/demo app and run all activities to avoid regressions
- [x] I have tested via a test drive, or a simulation/mock location app
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have updated the `CHANGELOG` including this PR
- [x] We might need to update / push `api/current.txt` files after running `$> make 1.0-core-update-api` (Core) / `$> make 1.0-ui-update-api` (UI) if there are changes / errors we're 🆗 with (e.g. `AddedMethod` changes are marked as errors but don't break SemVer) 🚀 If there are SemVer breaking changes add the `SEMVER` label. See [Metalava](https://github.com/mapbox/mapbox-navigation-android/blob/master/docs/metalava.md) docs

### CODE REVIEW

File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/instruction/InstructionView.java [hunk 1]
```diff
@@ -680,41 +650,20 @@ private void updateBannerInstructions(BannerText primaryBanner, BannerText secon
     }
   }
 
-  private void showGuidanceView(float currentStepDistanceTraveled, double currentStepTotalDistance) {
-    int metersBeforeShowGuidanceView = 200;
-    if (currentStepTotalDistance <= metersBeforeShowGuidanceView) {
```
Comment (LukasPaczos) [inline libnavigation-ui/src/main/java/com/mapbox/navigation/ui/instruction/InstructionView.java:685]: Are all of the step distance checks not needed anymore? What has changed?

File: libnavigator/src/main/java/com/mapbox/navigation/navigator/internal/MapboxNativeNavigatorImpl.kt [hunk 2]
```diff
@@ -397,9 +397,6 @@ object MapboxNativeNavigatorImpl : MapboxNativeNavigator {
                             }
                             routeProgressBuilder.bannerInstructions(bannerInstructions)
                         }
-                        ifNonNull(currentStep.bannerInstructions()) {
-                            stepProgressBuilder.guidanceViewURL(getGuidanceViewUrl(it))
```
Comment (LukasPaczos) [inline libnavigator/src/main/java/com/mapbox/navigation/navigator/internal/MapboxNativeNavigatorImpl.kt:401]: Does moving the url back to banner instructions allow us to omit step calculations?

File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/instruction/InstructionView.java (hunk 1 above)
Comment (abhishek1508) [reply libnavigation-ui/src/main/java/com/mapbox/navigation/ui/instruction/InstructionView.java:685]: Yeah banner instructions tell us when to display the guidance view. We were doing unnecessary calculations earlier. These are definitely not needed. The implementation with this PR is much more promising.
cc @LukasPaczos

File: libnavigator/src/main/java/com/mapbox/navigation/navigator/internal/MapboxNativeNavigatorImpl.kt (hunk 2 above)
Comment (abhishek1508) [reply libnavigator/src/main/java/com/mapbox/navigation/navigator/internal/MapboxNativeNavigatorImpl.kt:401]: Yes as mentioned above, these were unnecessary. We have all the information in banner instructions.
cc @LukasPaczos

File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/instruction/InstructionView.java [hunk 3]
```diff
@@ -195,16 +184,11 @@ public void subscribe(LifecycleOwner owner, NavigationViewModel navigationViewMo
     lifecycleOwner.getLifecycle().addObserver(this);
     this.navigationViewModel = navigationViewModel;
 
-    navigationViewModel.retrieveRouteProgress().observe(lifecycleOwner, routeProgress -> {
-      if (routeProgress != null) {
-        updateDistanceWith(routeProgress);
```
Comment (JunDai) [inline libnavigation-ui/src/main/java/com/mapbox/navigation/ui/instruction/InstructionView.java:200]: this **CAN NOT** be removed. The InsturctionView has two updates:
* `routeProgress` is for the distance text and the InstructionList update
* `bannerInstructions` below is for maneuver icon and the top banner instruction update

Comment (abhishek1508) [reply libnavigation-ui/src/main/java/com/mapbox/navigation/ui/instruction/InstructionView.java:200]: Good catch. Missed it. Fixed now

File: examples/src/main/java/com/mapbox/navigation/examples/ui/NavigationViewFragment.kt [hunk 4]
```diff
@@ -45,8 +44,7 @@ class NavigationViewFragment : Fragment(), OnNavigationReadyCallback, Navigation
         activity?.let {
             navigationView.initialize(
                 this,
-                getInitialCameraPosition(),
-                Utils.getMapboxAccessToken(it)
```
Comment (kmadsen) [inline examples/src/main/java/com/mapbox/navigation/examples/ui/NavigationViewFragment.kt:49]: 💯

File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/instruction/InstructionView.java [hunk 5]
```diff
@@ -410,32 +400,18 @@ public NavigationAlertView retrieveAlertView() {
...
-   * @param routeProgress {@link RouteProgress}
+   * @param bannerInstructions {@link BannerInstructions}
    */
-  public void determineGuidanceView(RouteProgress routeProgress) {
-    if (routeProgress != null && !isRerouting) {
-      double currentStepTotalDistance = 0.0;
-      float currentStepDistanceTraveled = 0f;
-      String guidanceUrl = null;
-
-      RouteLegProgress legProgress = routeProgress.getCurrentLegProgress();
```
Comment (kmadsen) [inline libnavigation-ui/src/main/java/com/mapbox/navigation/ui/instruction/InstructionView.java:421]: delete ❤️

File: examples/src/main/java/com/mapbox/navigation/examples/ui/CustomCameraActivity.kt [hunk 6]
```diff
@@ -18,7 +17,7 @@ import com.mapbox.navigation.ui.camera.SimpleCamera
 import com.mapbox.navigation.ui.listeners.BannerInstructionsListener
 import com.mapbox.navigation.ui.listeners.NavigationListener
 import com.mapbox.navigation.ui.map.NavigationMapboxMap
-import kotlinx.android.synthetic.main.activity_navigation_view.navigationView
+import kotlinx.android.synthetic.main.activity_navigation_view.*
```
Comment (JunDai) [inline examples/src/main/java/com/mapbox/navigation/examples/ui/CustomCameraActivity.kt:20]: nit: unnecessary change?

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/trip/session/MapboxTripSession.kt [hunk 7]
```diff
@@ -64,7 +65,8 @@ class MapboxTripSession(
     private val navigatorPredictionMillis: Long,
     private val navigator: MapboxNativeNavigator = MapboxNativeNavigatorImpl,
     threadController: ThreadController = ThreadController,
-    private val logger: Logger
+    private val logger: Logger,
+    private val accessToken: String?
```
Comment (JunDai) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/trip/session/MapboxTripSession.kt:69]: will it be possible to be **NULL**? I am asking this because I am thinking what the Line 477 will be when it's NULL and what will backend service return to us if we pass a null `access_token`?

File: libnavigation-core/src/test/java/com/mapbox/navigation/core/internal/trip/session/MapboxTripSessionTest.kt [hunk 8]
```diff
@@ -100,7 +100,8 @@ class MapboxTripSessionTest {
             locationEngine,
             navigatorPredictionMillis,
             navigator,
-            logger = logger
+            logger = logger,
+            accessToken = "pk.1234"
```
Comment (JunDai) [inline libnavigation-core/src/test/java/com/mapbox/navigation/core/internal/trip/session/MapboxTripSessionTest.kt:104]: same question as above in the `MapboxTripSession`.
If the `accessToken` is possible to be **NULL**, can we add a test case to cover the expected behavior here?

File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/instruction/InstructionView.java [hunk 9]
```diff
@@ -195,16 +184,17 @@ public void subscribe(LifecycleOwner owner, NavigationViewModel navigationViewMo
     lifecycleOwner.getLifecycle().addObserver(this);
     this.navigationViewModel = navigationViewModel;
 
+    navigationViewModel.retrieveBannerInstructions().observe(lifecycleOwner, bannerInstructions -> {
+      toggleGuidanceView(bannerInstructions);
```
Comment (JunDai) [inline libnavigation-ui/src/main/java/com/mapbox/navigation/ui/instruction/InstructionView.java:188]: do we need to expose this as a public api?
I am just curious can this logic be handled inside the InstructionView or maybe we could expose an option to let user turn on/off `GuidanceView`.
So we won't introduce another `SEMVER` change in near future to move the `GuidanceView` out of `InstructionView`.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/trip/session/MapboxTripSession.kt (hunk 7 above)
Comment (abhishek1508) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/trip/session/MapboxTripSession.kt:69]: The access token sent here is from `NavigationOptions` which has it nullable. Having null accessToken in this case is ok because if we send null in line 477, the user won't be able to draw the guidance view. In order to make it non-null I will have to change it in `NavigationOptions` and I don't want to do that, as I am not sure what kind of regressions this might introduce.

File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/instruction/InstructionView.java (hunk 9 above)
Comment (abhishek1508) [reply libnavigation-ui/src/main/java/com/mapbox/navigation/ui/instruction/InstructionView.java:188]: None of our customers use `InstructionView`. I will let it be as is for now, because I don't see why we would want to move it out as a separate view. All it is an image view. Customers have been drawing guidance views by themselves as they are available in banner instructions.

File: libnavigation-core/src/test/java/com/mapbox/navigation/core/internal/trip/session/MapboxTripSessionTest.kt (hunk 8 above)
Comment (abhishek1508) [reply libnavigation-core/src/test/java/com/mapbox/navigation/core/internal/trip/session/MapboxTripSessionTest.kt:104]: We cannot add a unit test case in case the token is NULL. This would be better covered with instrumentation test, as if the token is null, you would see no guidance view being drawn and we could have verify that. But, in this case there is nothing to verify from unit test point of view.

File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/instruction/InstructionView.java (hunk 9 above)
Comment (JunDai) [reply libnavigation-ui/src/main/java/com/mapbox/navigation/ui/instruction/InstructionView.java:188]: since we talked about this several days ago when we discussed the GuidanceView landscape UX and we were talking about moving the guidanceView logic out of InstructionView.
It's fine that not having an individual component just for the guidance view as you mentioned above it's just an image view.

But I think exposing this new public API in InstructionView seems weird. We may leave it here for now if you prefer. But we should think of having a separate GuidanceImageObserver, or sort of, in Core and the developer who wants guidanceImage can just register for that without knowing the InstructionView.
This way, we can solve the 1tap guidance view problem(https://github.com/mapbox/mapbox-navigation-android/issues/3303) too. As you said, 1tap can draw the image wherever they like. cc: @LukasPaczos

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/trip/session/MapboxTripSession.kt (hunk 7 above)
Comment (JunDai) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/trip/session/MapboxTripSession.kt:69]: It can be tracked as a follow up issue. I did a quick check about the usage of `access_token` in `NavigationOptions`. 
I think we do have the `NULL` check when we initiate the `router` module in `MapboxNavigation`
```kotlin
MapboxModuleType.NavigationOffboardRouter -> arrayOf(
                String::class.java to (accessToken
                    ?: throw RuntimeException(MAPBOX_NAVIGATION_TOKEN_EXCEPTION_OFFBOARD_ROUTER)),
                Context::class.java to navigationOptions.applicationContext,
                UrlSkuTokenProvider::class.java to MapboxNavigationAccounts.getInstance(navigationOptions.applicationContext)
            )
```

To me it seems the `access_token` in `NavigationOptions` should be a non-null property, isn't it?
cc: @Guardiola31337 , @kmadsen , @LukasPaczos

**Comment by JunDai** [comment]
btw, CI isn't happy :(
cc: @abhishek1508

File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/instruction/InstructionView.java (hunk 9 above)
Comment (abhishek1508) [reply libnavigation-ui/src/main/java/com/mapbox/navigation/ui/instruction/InstructionView.java:188]: Yes, and all 1TAP needs to do is the following without any dependency from UI SDK. Inside their `BannerInstructionObserver` 
```
val bannerView = bannerInstruction.view();
List<BannerComponents> components = bannerView.components();
    if (components != null) {
      String urlToLoad = components.get(0).imageUrl();
      new Picasso
          .Builder(getContext())
          .downloader(new OkHttp3Downloader(getClient()))
          .build()
          .load(urlToLoad)
          .into(guidanceViewImage);
    }
```
This way you don't need to use the GuidanceView from `InstructionView`, but if you are happy with Mapbox's implementation, you can just invoke the API as provided.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/trip/session/MapboxTripSession.kt (hunk 7 above)
Comment (abhishek1508) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/trip/session/MapboxTripSession.kt:69]: I totally agree that it should be NON-NULL, but I think I am missing the discussion as to why was it made `NULLABLE` in the first place.

Comment (LukasPaczos) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/trip/session/MapboxTripSession.kt:69]: The only places in the code that use the token are the routers. Since they are modular, the developer could replace the implementations and not actually need the token, that's why it's nullable.

File: libnavigation-core/src/test/java/com/mapbox/navigation/core/internal/trip/session/MapboxTripSessionTest.kt (hunk 8 above)
Comment (LukasPaczos) [reply libnavigation-core/src/test/java/com/mapbox/navigation/core/internal/trip/session/MapboxTripSessionTest.kt:104]: We could verify the URL structure without a token.

Comment (LukasPaczos) [reply libnavigation-core/src/test/java/com/mapbox/navigation/core/internal/trip/session/MapboxTripSessionTest.kt:104]: Also would be great to add a test that verifies the behavior if there is a token.

File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/instruction/InstructionView.java (hunk 9 above)
Comment (LukasPaczos) [reply libnavigation-ui/src/main/java/com/mapbox/navigation/ui/instruction/InstructionView.java:188]: > This way you don't need to use the GuidanceView from InstructionView, but if you are happy with Mapbox's implementation, you can just invoke the API as provided.

Yea, I think this is flexible enough, we shouldn't need another observer just for the guidance view type :+1: 

There is one caveat though, the user agent has to contain `MapboxJava`. We should document it in the `BannerInstruction` object for any developers that would like to make the request themselves. Unless this is changing soon @danpaz?

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt [hunk 10]
```diff
@@ -451,8 +453,20 @@ internal class MapboxTripSession(
...
-                action(it)
+            ifNonNull(bannerInstructionEvent.bannerInstructions) { bannerInstructions ->
+                val bannerView = bannerInstructions.view()
+                val bannerComponents = bannerView?.components()
+                ifNonNull(bannerComponents) { components ->
+                    components.forEachIndexed { index, component ->
+                        component.takeIf { it.type() == BannerComponents.GUIDANCE_VIEW }?.let { c ->
+                            components[index] =
+                                c.toBuilder()
+                                    .imageUrl(c.imageUrl()?.plus("&access_token=$accessToken"))
```
Comment (LukasPaczos) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt:464]: We're missing a test case for this.

Comment (LukasPaczos) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt:464]: Hmm, still not seeing the test, but you've mentioned we already have one @abhishek1508?
